packages/@lwc/template-compiler/src/config.ts

```diff
@@ -26,20 +34,62 @@ export interface Config {
     preserveHtmlComments?: boolean;
 }
 
-export type NormalizedConfig = Required<Config>;
+export type NormalizedConfig = Required<Omit<Config, 'customRendererConfig'>> &
+    Partial<Pick<Config, 'customRendererConfig'>>;
 
 const AVAILABLE_OPTION_NAMES = new Set([
+    'customRendererConfig',
     'experimentalComputedMemberExpression',
     'experimentalDynamicDirective',
     'preserveHtmlComments',
 ]);
 
+function normalizeCustomRendererConfig(config: CustomRendererConfig): CustomRendererConfig {
+    const tagNames: string[] = [];
+    const normalizedConfig: CustomRendererConfig = {
+        elements: config.elements.map((e) => {
+            const tagName = e.tagName.toLowerCase();
+            // Custom element cannot be allowed to have a custom renderer hook
+            // The renderer is cascaded down from the owner(custom element) to all its child nodes who
+            // do not have a renderer specified.
+            invariant(!isCustomElementTag(tagName), TemplateErrors.CUSTOM_ELEMENT_TAG_DISALLOWED, [
+                e.tagName,
+            ]);
+
+            tagNames.push(tagName);
+            return {
+                tagName,
+                namespace: e.namespace?.toLowerCase(),
+                attributes: e.attributes?.map((a) => a.toLowerCase()),
+            };
+        }),
+        directives: config.directives.map((d) => d.toLowerCase()),
+    };
```

Comment:
I could have sworn we had decided it was better to expose a function here rather than a declarative config object? That way, it's easier for custom renderer implementers to change their implementation to have complex logic (e.g. "this tag, but only if it has this attribute...").

I'm also not sure we should be in the business of checking whether a custom element tag is used here. If the implementer of a custom renderer makes a mistake and tries to use a custom renderer for a custom element, and it fails at runtime, shouldn't that be their responsibility?

Reply:
Yes, we did discuss about a function v/s declarative config, the decision was to go with a declarative config. 

- The problem with function is that it would be driven by the implemented of custom renderer and what they need to make a decision. It's adds a maintenance burden.
- To retain the decision to qualify an element for custom renderer internal and also keep it common regardless of the implementers, it was decided to go with declarative config.

Covered in the [design doc](https://salesforce.quip.com/zD0PAQbahVk0#temp:C:GJD2d710bac48d5422aaa456d049).

> checking whether a custom element tag
For background, this is a side effect of https://github.com/salesforce/lwc/pull/2763#issuecomment-1145173224.
This is more of a limitation of the current design than a restriction on the implementer. The owner's renderer is [cascaded down](https://github.com/salesforce/lwc/pull/2763/files#diff-a9673af211cdbdb5096c7ee4533b234a8bf90e9271341048b42ffcd6c0e143ceR247) to its children if they do not have a custom renderer specified.
```xml
<template>
     <x-foo class='custom-renderer'> </x-foo>
</template>
```
if `x-foo` is qualified to get a custom renderer by its parent, x-foo will get it. Since the renderer is propagated down to all nodes of x-foo via {vnode.ownner to all elements in x-foo's template, they will inherit the custom renderer.